### ISSUE TITLE
[jit][script] literals don't interact well with gpu tensors

### ISSUE THREAD

**Issue by zdevito** [pytorch/pytorch#7504]
Literals are always allocated on the CPU, meaning that simple expressions break when the input tensors are on the GPU, e.g.:

```
@torch.jit.script
def foo(a):
    return 1 + a # fails for GPU tensor
```
If we added types for non-tensors numbers then we would be able to patch the graph to ensure things would work, e.g:

```
one = to_backend_as(1, a)
return one + a
```

Alternatively, we can scalar tensors like 1 have this behavior universally in the operator library.

Steps to resolve this issue:
1. Introduce `float` and `int` types into type.h, and change the frontend to assign these types to constant literals like `1`.
2. Modify emitSimpleExpr to allow float/int to work on normal arithmetic operators like +, and ensure the returned value has the float/int type as well.
2. Modify `emitSimpleExpr` and `tryMatchSchema` in compiler.cpp to add an implicit conversion from int/float into a tensor type (see "// check input types" for where this occurs).  Given an `int` or `float` type, this implicit conversion should insert code that converts the float/int value `v` into a tensor on the correct backend for the op by finding the first Tensor argument to the op `t` and then inserting a call to move `v` to the same device as `t`.

**Comment by apaszke** [comment]
We should just allow for type conversions on scalars. The `to_backend_as` could be a temporary solution, but I think we should remove it before the JIT goes public.

**Comment by driazati** [comment]
Fixed in #8687 (followup to #8177)
